### ISSUE TITLE
[feat] Consistent dataset re-sampling

### ISSUE THREAD

**Issue by jlamypoirier** [ServiceNow/Fast-LLM#130]
# 🧐 Problem Description

We often want to restart/resume an experiment with extra training samples, but currently this causes datasets to be re-sampled in an entirely different way, which means for example we're going to reuse lots of sample when resuming mid-epoch instead of using the rest , so training may be sub-optimal. We want to address that. 

# 💡 Proposed Solution

The main consistency issue is with shuffling , so we'll want to address that:
* Instead of shuffling all epochs together, add an option to shuffle epoch independently.
* Shuffle the last epoch as a whole instead of just the samples we want.
* Add an option to disable shuffling entirely. This will be especially useful for sampling << 1 epoch, for which shuffling is unnecessary and significantly increases sampling time.

**Comment by tscholak** [comment]
Hi @jlamypoirier, training data these days is never really just one dataset. It's many different datasets, some repeated many times over, others barely reaching one epoch. Batches should never be one dataset or another, it should always be sampled according to predefined sampling weights.

**Comment by jlamypoirier** [comment]
I perfectly agree. But this refers to individual dataset sampling which comes before and independently from the blending and batching. Blending and batching are already behaving as you're describing and this won't change.
